pkg/genericapiserver/api/handlers/rest.go

```diff
@@ -1079,22 +1113,95 @@ func setListSelfLink(obj runtime.Object, req *restful.Request, namer ScopeNamer)
 	return count, err
 }
 
-func getPatchedJS(patchType types.PatchType, originalJS, patchJS []byte, obj runtime.Object) ([]byte, error) {
+func patchObjectJSON(
+	patchType types.PatchType,
+	codec runtime.Codec,
+	originalObject runtime.Object,
+	patchJS []byte,
+	objToUpdate runtime.Object,
+	versionedObj runtime.Object,
+) (originalObjJS []byte, patchedObjJS []byte, retErr error) {
+	if js, err := runtime.Encode(codec, originalObject); err != nil {
+		return nil, nil, err
+	} else {
+		originalObjJS = js
+	}
 	switch patchType {
 	case types.JSONPatchType:
 		patchObj, err := jsonpatch.DecodePatch(patchJS)
 		if err != nil {
-			return nil, err
+			return nil, nil, err
+		}
+		if patchedObjJS, err = patchObj.Apply(originalObjJS); err != nil {
+			return nil, nil, err
 		}
-		return patchObj.Apply(originalJS)
 	case types.MergePatchType:
-		return jsonpatch.MergePatch(originalJS, patchJS)
+		var err error
+		if patchedObjJS, err = jsonpatch.MergePatch(originalObjJS, patchJS); err != nil {
+			return nil, nil, err
+		}
 	case types.StrategicMergePatchType:
-		return strategicpatch.StrategicMergePatchData(originalJS, patchJS, obj)
+		var err error
+		if patchedObjJS, err = strategicpatch.StrategicMergePatch(originalObjJS, patchJS, versionedObj); err != nil {
+			return nil, nil, err
+		}
 	default:
 		// only here as a safety net - go-restful filters content-type
-		return nil, fmt.Errorf("unknown Content-Type header for patch: %v", patchType)
+		return nil, nil, fmt.Errorf("unknown Content-Type header for patch: %v", patchType)
+	}
+	if err := runtime.DecodeInto(codec, patchedObjJS, objToUpdate); err != nil {
+		return nil, nil, err
+	}
+	return
+}
+
+func strategicPatchObject(
```

Comment:
this style of arguments is somewhat nonstandard and hard to read - if we don't think we're going to have fewer arguments here in the future, it might be more appropriate to create a struct that holds codec, originalObject, and a few others and have methods on it like `Patch(patch []byte, objToUpdate)` or something similar.  You don't have to do it here, but this entire package is somewhat hard to read as is.